src/protocol/libp2p/kademlia/mod.rs

```diff
@@ -485,11 +485,20 @@ impl Kademlia {
 
                         // update routing table and inform user about the update
                         self.update_routing_table(&peers).await;
-                        self.engine.register_response(
-                            query_id,
-                            peer,
-                            KademliaMessage::GetRecord { key, record, peers },
-                        );
+
+                        // Send partial response to the user.
+                        if let Some(QueryAction::GetRecordPartialResult { query_id, record }) =
+                            self.engine.register_response(
+                                query_id,
+                                peer,
+                                KademliaMessage::GetRecord { key, record, peers },
+                            )
+                        {
+                            let _ = self
+                                .event_tx
+                                .send(KademliaEvent::GetRecordPartialResult { query_id, record })
+                                .await;
+                        }
```

Comment:
nit: to keep the structure of event handling, I would let the `engine` to pass `QueryAction::GetRecordPartialResult` to `on_query_action()` and published the outbound event there. The latency won't be noticeable, but the code structure will be more streamlined.

Reply:
Thanks, that makes sense! 🙏 Let me know if the latest changes are oki with you, planning to do a release later today